src/semant/Semant.cpp

```diff
@@ -1605,13 +1605,49 @@ void Semant::transEvidence(absyn::Evidence* ne) {
 }
 
 void Semant::transQuery(absyn::Query* nq) {
+  // special check for for-loop in query statement
+  std::vector<std::shared_ptr<ir::VarDecl> > vds;
+  std::shared_ptr<ir::Expr> cond = nullptr;
+  if (nq->getVarDecls().size() > 0) { // has for loops
+    vds = transVarDecls(nq->getVarDecls(), true);
+    // add local variables
+    for (auto&v : vds) add_local_var(v);
```

Comment:
Okay, was thinking of something along these lines:
    `for (auto&v : vds) {`
        `// ensure that variables have fixed dimensions`
        `auto ty = dynamic_cast<const ir::NameTy*>(v->getTyp());`
        `assert(ty != NULL);`
        `if (ty->getRefer()->getPreLen() < 1) {`
            `error(nq->line, nq->col, "for-loop variable dimensions must be fixed!");`
        `}`
       `add_local_var(v);`
  `}`
but perhaps can be a little simpler.